File: q11.c
#include <stdio.h>

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j <= high - 1; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return i + 1;
}

void quick_sort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quick_sort(arr, low, pi - 1);
        quick_sort(arr, pi + 1, high);
    }
}

void find_neighbors_of_median(int arr[], int size) {
    // Sort the array
    quick_sort(arr, 0, size - 1);

    // Find the median index
    int median_index = size / 2;
    int median = arr[median_index];

    // Find the neighbors of the median
    int left_neighbor = (median_index > 0) ? arr[median_index - 1] : -1;
    int right_neighbor = (median_index < size - 1) ? arr[median_index + 1] : -1;

    printf("Left Neighbor: %d\n", left_neighbor);
    printf("Median: %d\n", median);
    printf("Right Neighbor: %d\n", right_neighbor);
}

int main() {
    int arr[] = {3, 1, 6, 2, 9, 8, 4, 5, 7};
    int size = sizeof(arr) / sizeof(arr[0]);

    find_neighbors_of_median(arr, size);

    return 0;
}

File: q11.py
def partition(arr, low, high):
    pivot = arr[high]  # Choosing the last element as the pivot
    i = low - 1  # Index of smaller element

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def find_median(arr):
    arr.sort()
    median_index = len(arr) // 2
    median = arr[median_index]
    return median, median_index

def find_neighbors_of_median(arr):
    median, median_index = find_median(arr)
    partition(arr, 0, len(arr) - 1)
    left_neighbor = arr[median_index - 1] if median_index > 0 else None
    right_neighbor = arr[median_index + 1] if median_index < len(arr) - 1 else None
    return left_neighbor, median, right_neighbor

# Example usage:
array = [3, 1, 6, 2, 9, 8, 4, 5, 7]
left_neighbor, median, right_neighbor = find_neighbors_of_median(array)
print("Left Neighbor:", left_neighbor)
print("Median:", median)
print("Right Neighbor:", right_neighbor)
